embassy-usb-driver/src/host.rs

```diff
@@ -0,0 +1,309 @@
+    ///
+    /// This can be a scarce resource, for one-off requests please scope the channel so it's dropped after completion
+    ///
+    /// `pre` - device is low-speed and communication is going through hub, so send PRE packet
+    fn alloc_channel<T: channel::Type, D: channel::Direction>(
+        &self,
+        addr: u8,
+        endpoint: &EndpointInfo,
+        pre: bool,
+    ) -> Result<Self::Channel<T, D>, HostError>;
+
+    // Drop happens implicitly on channel-side
+    // / Drop allocated channel
+    // fn drop_channel<T: channel::Type, D: channel::Direction>(&self, channel: &mut Self::Channel<T, D>);
+}
+
+/// [UsbChannel] Typelevel structs and traits
+// TODO: Seal traits
+pub mod channel {
+    use super::EndpointType;
+
+    pub trait Type {
+        fn ep_type() -> EndpointType;
+    }
+    pub struct Control {}
+    pub struct Interrupt {}
+    pub struct Bulk {}
+    pub struct Isochronous {}
+    impl Type for Control {
+        fn ep_type() -> EndpointType {
+            EndpointType::Control
+        }
+    }
+    impl Type for Interrupt {
+        fn ep_type() -> EndpointType {
+            EndpointType::Interrupt
+        }
+    }
+    impl Type for Bulk {
+        fn ep_type() -> EndpointType {
+            EndpointType::Bulk
+        }
+    }
+    impl Type for Isochronous {
+        fn ep_type() -> EndpointType {
+            EndpointType::Isochronous
+        }
+    }
+
+    #[diagnostic::on_unimplemented(message = "This is not a CONTROL channel")]
+    pub trait IsControl {}
+    impl IsControl for Control {}
+
+    #[diagnostic::on_unimplemented(message = "This is not a CONTROL channel")]
+    pub trait IsInterrupt {}
+    impl IsInterrupt for Interrupt {}
+
+    pub trait Direction {
+        fn is_in() -> bool;
+        fn is_out() -> bool;
+    }
+    pub struct In {}
+    pub struct Out {}
+    pub struct InOut {}
+    impl Direction for In {
+        fn is_in() -> bool {
+            true
+        }
+        fn is_out() -> bool {
+            false
+        }
+    }
+    impl Direction for Out {
+        fn is_in() -> bool {
+            false
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+    impl Direction for InOut {
+        fn is_in() -> bool {
+            true
+        }
+        fn is_out() -> bool {
+            true
+        }
+    }
+
+    #[diagnostic::on_unimplemented(message = "This is not an IN channel")]
+    pub trait IsIn: Direction {}
+    impl IsIn for In {}
+    impl IsIn for InOut {}
+
+    #[diagnostic::on_unimplemented(message = "This is not an OUT channel")]
+    pub trait IsOut: Direction {}
+    impl IsOut for Out {}
+    impl IsOut for InOut {}
+}
+
+/// Specify the timeout of a channel
+pub struct TimeoutConfig {
+    /// Maximum response timeout for transactions with a Data Stage
+    pub data_timeout: Duration,
+    /// Maximum response timeout for transactions without a data stage
+    pub standard_timeout: Duration,
+}
```

Comment:
Could these be substituted with `with_timeout` on the user side? These don't seem granular enough to be useful.

Reply:
The main issue with `with_timeout` is that it has to be done for each request, if forgotten the request may stall infinitely. And since the standard already provides timeout defaults/recommendations, it makes more sense to configure it up-front (with the spec's recommended values) rather than wrap each call that is a transfer.

Reply:
That's true. Do those timeouts only apply to control transfers?

Reply:
I believe the spec doesn't differentiate between types for timeouts just if it has a data stage or not (with the exception of SetAddress command). So the timeouts apply to all transfers on a specific channel.

Reply:
Interrupt transfers should have a timeout (for the same reasons as before), as they *need* to be nak'ed for it to be a valid poll (they're not actually interrupts). Isochronous makes sense to omit.

You could specify that the timeout is specifically for the data stage of a request as is in the standard (which is valid for interrupt & bulk), however looking at the dwc driver it's already not possible to make this distinction with the hw.

We could do `set_timeout(Duration)` & `set_timeout_control(Duration, Duration`  & missing for isochronous; which doesn't seem as nice. If you have any ideas for this I'm open to suggestions.